### ISSUE TITLE
async-std@1.6-beta.1: panicked at `async fn` resumed after completion

### ISSUE THREAD

**Issue by arbfranklin** [async-rs/async-std#777]
I have some code that automatically attempts to reconnect on connection failure. It runs ok in 1.5.0, but panics on 1.6-beta.

Issue seems to be on my side, I wrap a Future which after a failed reconnect resolves to a Poll:Ready(Err).. and then when I poll it again, it panics... however, the prior runtime (and tokio) seem to be ok with it.

Here are parts of a trace that might be of interest:

```
thread 'main' panicked at '`async fn` resumed after completion', <... snip ...>
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /Users/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.44/src/backtrace/libunwind.rs:86
...
  15: <std::future::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libstd/future.rs:44
  16: <core::pin::Pin<P> as core::future::future::Future>::poll
             at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libcore/future/future.rs:118
  17: futures_util::future::future::FutureExt::poll_unpin
             at /Users/a/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.4/src/future/future/mod.rs:507
... snip user code ...
  19: <futures_util::stream::stream::split::SplitStream<S> as futures_core::stream::Stream>::poll_next
             at /Users/a/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.4/src/stream/stream/split.rs:31
  20: <futures_util::stream::stream::filter_map::FilterMap<St,Fut,F> as futures_core::stream::Stream>::poll_next
             at /Users/a/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.4/src/stream/stream/filter_map.rs:106
  21: <S as futures_core::stream::TryStream>::try_poll_next
             at /Users/a/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-core-0.3.4/src/stream.rs:193
  22: <futures_util::stream::try_stream::map_err::MapErr<St,F> as futures_core::stream::Stream>::poll_next
             at /Users/a/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.4/src/stream/try_stream/map_err.rs:92
  23: <core::pin::Pin<P> as futures_core::stream::Stream>::poll_next
             at /Users/a/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-core-0.3.4/src/stream.rs:121
  24: futures_core::stream::if_alloc::<impl futures_core::stream::Stream for alloc::boxed::Box<S>>::poll_next
             at /Users/a/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-core-0.3.4/src/stream.rs:209
  25: futures_util::stream::stream::StreamExt::poll_next_unpin
             at /Users/a/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.4/src/stream/stream/mod.rs:1184
  26: <futures_util::stream::stream::next::Next<St> as core::future::future::Future>::poll
             at /Users/a/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.4/src/stream/stream/next.rs:35
  27: std::future::poll_with_tls_context
             at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libstd/future.rs:102
... snip user code ...
  29: <std::future::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libstd/future.rs:44
  30: <async_std::task::builder::SupportTaskLocals<F> as core::future::future::Future>::poll::{{closure}}
             at /Users/a/.cargo/registry/src/github.com-1ecc6299db9ec823/async-std-1.6.0-beta.1/src/task/builder.rs:174
  31: async_std::task::task_locals_wrapper::TaskLocalsWrapper::set_current::{{closure}}
             at /Users/a/.cargo/registry/src/github.com-1ecc6299db9ec823/async-std-1.6.0-beta.1/src/task/task_locals_wrapper.rs:60
  32: std::thread::local::LocalKey<T>::try_with
             at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libstd/thread/local.rs:262
  33: std::thread::local::LocalKey<T>::with
             at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libstd/thread/local.rs:239
  34: async_std::task::task_locals_wrapper::TaskLocalsWrapper::set_current
             at /Users/a/.cargo/registry/src/github.com-1ecc6299db9ec823/async-std-1.6.0-beta.1/src/task/task_locals_wrapper.rs:55
  35: <async_std::task::builder::SupportTaskLocals<F> as core::future::future::Future>::poll
             at /Users/a/.cargo/registry/src/github.com-1ecc6299db9ec823/async-std-1.6.0-beta.1/src/task/builder.rs:172
  36: smol::run::run::{{closure}}::{{closure}}
             at /Users/a/.cargo/registry/src/github.com-1ecc6299db9ec823/smol-0.1.4/src/run.rs:143
  37: scoped_tls_hkt::ScopedKey<T>::set
             at /Users/a/.cargo/registry/src/github.com-1ecc6299db9ec823/scoped-tls-hkt-0.1.2/src/lib.rs:488
  38: smol::throttle::setup
             at /Users/a/.cargo/registry/src/github.com-1ecc6299db9ec823/smol-0.1.4/src/throttle.rs:31
```

**Comment by arbfranklin** [comment]
Root cause seems to be continuing to poll a future that has previously resolved to an error -- obviously incorrect, however, the prior runtime (and tokio) seem to be ok with it.

**Comment by dignifiedquire** [comment]
@stjepang any thoughts, if this is on purpose or not?

**Comment by stjepang** [comment]
I cannot figure out from the backtrace what is wrong or how the panic is related to async-std/smol. There are some frames replaced with "..."

Some future created by an async {} block is poller after it has been completed. However, judging from the stacktrace, this future seems deep inside user code and unrelated to the runtime.

@arbfranklin Could you perhaps provide a few more of the missing frames? We should find out what that future is.

**Comment by arbfranklin** [comment]
Seems to be bad error propagation on my part... I'm trying to catch io::ErrorKind::ConnectionRefused for when a port is unavailable, and that seems to have changed in 1.6?

```
use async_std::net::TcpStream;
use std::net::SocketAddr;

#[async_std::main]
async fn main() {
    let addr: SocketAddr = "127.0.0.1:12345".parse().unwrap();
    TcpStream::connect(addr).await.unwrap();
}
```

Output on 1.5.0:

```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 61, kind: ConnectionRefused, message: "Connection refused" }', examples/async_panic.rs:7:5
```

Output on 1.6.0-beta.2:

```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 32, kind: BrokenPipe, message: "Broken pipe" }', examples/async_panic.rs:7:5
```

_Note: I'm using macOs Mojave_

**Comment by stjepang** [comment]
Looks like an issue in smol, wrong error kind is being returned if connection is refused.

**Comment by stjepang** [comment]
Published a new version. Can you confirm the issue is fixed?

**Comment by arbfranklin** [comment]
Thanks @stjepang, I have confirmed that the issue is resolved with smol 0.1.10 / async-std 1.6.0.
